### ISSUE TITLE
✨ Add focus decoration to UI elements when doing keyboard navigation

### ISSUE THREAD

**Issue by goanpeca** [napari/napari#2999]
## 🚀 Feature

See https://github.com/napari/napari/issues/2993

## Motivation

When doing tab/keyboard only navigation in napari, the elements that have the focus should be highlighted. Options include color changes or border changes.

**Comment by goanpeca** [comment]
@liaprins I opened this new issue to track the focus discussion:

From @liaprins at https://github.com/napari/napari/issues/2993#issuecomment-877313089
>Yes, I agree that showing an in-focus state would be helpful, ideally if it can align with the OS's default design ... is it possible to do this with QT @goanpeca ? If we cannot call on the existing OS's focus design, then I think in general using a blue outline (similar to Mac OS treatment for in-focus) would communicate the idea.

>Lastly, I was not sure what you meant by this @goanpeca :

>>Mac by default disables tab navigation and focussing

>Do you mean that happens in certain cicrumstance that happens to the Preferences dialog? If so, what causes that? Ideally it would not be disabling tab navigating and focus states (for a11y purposes)?

**Comment by goanpeca** [comment]
Thanks for chiming in @liaprins. To your questions:

>Yes, I agree that showing an in-focus state would be helpful, ideally if it can align with the OS's default design ... is it possible to do this with QT @goanpeca ? If

Yes, for some controls, for others it requires a bit more work (need code changes and not just QSS changes), but doable.

>If we cannot call on the existing OS's focus design, then I think in general using a blue outline (similar to Mac OS treatment for in-focus) would communicate the idea.

Indeed. 

>Mac by default disables tab navigation and focussing

![image](https://user-images.githubusercontent.com/3627835/125110807-b1623e80-e0aa-11eb-94ae-d5046337fa84.png)

Mac by default has no tab navigation for all elements, on Qt we can choose to respect this, or always force tab naviagtion of all elements, even if that OSX setting above is not enabled on the preferences. I have always found that setting puzzling.

I would like to make napari as usable as possible without a mouse (yes for a UI dealing with graphics feels daunting).

**Comment by liaprins-czi** [comment]
### Overview of design need
- **Also link to any specific comments or threads where the problem to be solved by design is mentioned**
https://github.com/napari/napari/issues/2993#issuecomment-877313089

- **In a sentence or two, describe the problem to be solved for users**
Make keyboard navigation visually accessible.

### What level of design is needed? (Choose all that apply)
- [X] **User experience research:**  high-level recommendation/exploration of user needs, design heuristics, and / or best practices to inform a design experience  
(Use this option when you feel there’s a challenge to be solved, but you’re curious about what the experience should be — may involve research studies to understand challenges/opportunities for design)
- [ ] **Information flow / conceptual:** organizing and structuring of information flow and content, including layout on screen or across multiple steps
- [X] **Visual:** creating mockups, icons, etc
(If choosing this level alone, it means that the content to be mocked up and its organization is already known and specified)

### Is design a blocker?
- [ ] **Yes:** engineering cannot proceed without a design first
- [X] **No:** engineering can create a first version, and design can come in later to iterate and refine

If selecting **Yes**, how much design input is needed to unblock engineering? For example, is a full, final visual design needed, or just a recommendation of which conceptual direction to go?

**Comment by isabela-pf** [comment]
Hi all! This is a discussion I love to see happening and hope to help move forward. Fair warning, this might be a little long because I’m trying to process the details I’m learning about focus states as I go.

## Current state
Based on my own fiddling around with napari, it seems like you can at least somewhat navigate through napari’s interface with a keyboard alone (an easy but by no means the only way to find focus quickly), it’s just a focus state that’s completely missing. I’m mostly able to verify this because I skimmed the interface with a screen reader which adds its own focus state on top.

## What napari needs
Focus—essentially where a user is immediately interacting with an interface—needs visual representation so that users can situate themself in an interface. This has been compared to the way a cursor gives feedback in a document or a mouse pointer everywhere else.

If you’re looking for standards, WCAG requires visible keyboard focus ([focus visible in the WCAG 2.2 working draft](https://www.w3.org/TR/WCAG22/#focus-visible)).

I took the time to read this really informative [guide to designing focus indicators](https://www.sarasoueidan.com/blog/focus-indicators/) if you are curious about this as well. I also tried tabbing around on Mac with and without a screen reader to see how this is being handled in a few applications on one OS.

Because there’s many types of content in napari, we are going to be sure that we not only have focus indicators that can be activated everywhere that is interactive, but also that the styling works for the variety of content and can be easily extended to new content.

## Proposals

I think the simplest approach would be to have an outline around different elements. This is kind of like the defaults I found when stumbling about Mac preferences and Firefox and Chrome, but per the blog post I linked I think we should add some spacing between the outline and the element. 

Starting with some of the basic interface interactions in the (default) dark theme:

### Option 1
Color: napari’s main UI blue
Width: 2px
Distance from UI element: 2px

An outline using the blue I’ve found in the rest of the interface. I do want to note that this blue doesn’t pass contrast on the dark backgrounds (it does on light though), but if we want a consistent interface and to change contrast on a single blue all at once later on, this is an option.

<img width="690" alt="Input Field with blue focus" src="https://user-images.githubusercontent.com/50221806/132936050-b634bfa9-59a9-46fa-906f-e1510111193a.png">
<img width="270" alt="napari layer list tab with blue focus" src="https://user-images.githubusercontent.com/50221806/132936055-be919f7d-9a23-469c-b7b9-b4651a3f18df.png">
<img width="950" alt="napari Preferences dialog with multiple blue focuses" src="https://user-images.githubusercontent.com/50221806/132936059-1df9277a-cc89-4ea4-be98-5b36aef5d60f.png">

We could also have this approach with an added blue shadow, similar to the Mac OS @liaprins-czi mentioned above. Since napari doesn’t have many gradients as far as I have seen, this doesn’t seem particularly necessary to me though.

<img width="690" alt="Input Field with blue focus" src="https://user-images.githubusercontent.com/50221806/132936069-c9eaedae-78f6-484f-ab5c-bf2e0ddb4c9d.png">
<img width="270" alt="napari layer list tab with blue focus" src="https://user-images.githubusercontent.com/50221806/132936070-5a0afbfc-0a4e-4eb6-889d-16a5b55e1773.png">
<img width="950" alt="napari Preferences dialog with multiple blue focuses" src="https://user-images.githubusercontent.com/50221806/132936071-82263367-9d5e-4d7b-9c1d-dfd5eaa11ba7.png">

### Option 2
Color: napari’s main UI blue and white
Width: 2px blue, 1px white
Distance from UI element: 2px

Some focus indicators use two contrasting side-by-side outlines to ensure contrast on varying backgrounds. Since napari can have variable content, especially in the viewer area or in extensions, this could be a good approach.

<img width="690" alt="Input Field with blue and white focus" src="https://user-images.githubusercontent.com/50221806/132936074-b2d0ff80-3896-4d46-a906-6a60c706ef93.png">
<img width="269" alt="napari layer list tab with blue and white focus" src="https://user-images.githubusercontent.com/50221806/132936075-e6bb1079-dbe0-4db9-b652-60c096043a4b.png">
<img width="950" alt="napari Preferences with multiple blue and white focuses" src="https://user-images.githubusercontent.com/50221806/132936080-cb71a3f2-0054-4a5e-b828-6f3204387580.png">

### Option 3
Color: napari’s main UI white and dark grey
Width: 2px white, 1px grey
Distance from UI element: 2px

Another set of contrasting outlines, but this time with the colors that already fill the UI. Because they aren’t using the typical highlight blue I think they could be harder to notice, but they also will have very good contrast without having to adjust the main UI blue.

In this case, the dark outline is on the dark background and is not visible until focus is over a light area.

<img width="690" alt="Input Field with white and grey focus" src="https://user-images.githubusercontent.com/50221806/132936084-361b2a53-4a5f-41fb-9bef-09920fb7f557.png">
<img width="269" alt="napari layer list tab with white and grey focus" src="https://user-images.githubusercontent.com/50221806/132936086-c57fb85f-7b42-42fc-8518-b557fe91039a.png">
<img width="950" alt="napari Preferences with multiple white and grey focuses" src="https://user-images.githubusercontent.com/50221806/132936088-5dd91568-fc6d-4cda-b317-ff6eb1e0361a.png">

## Things to think about
Both because of my limited experience and the many interaction types in napari, I think some experimentation will be needed as I’m not sure how this will fully play out. Some problems I anticipate:
- I will probably need help on solutions for UI elements that are pretty specific to napari and/or image-viewing software. I’m thinking about focus for things like points and shapes and so on where I have less points of reference.
- Users could have any content in the viewer area, so we likely need something that can have sufficient contrast on almost any background.
- Some elements—like buttons— are very close together and an outline may overlap with other elements around it.
- We may need per-theme color. The main blue I’m finding in napari’s interface doesn’t have enough contrast on many of the dark theme backgrounds, but does work on white. 
- Double-check elements automatically receive proper focus (for example, when dialogs appear they need to automatically gain focus). This might be better served as its own issue.

## Try it out yourself
I always encourage people to test things themself, and this is doubly true with accessibility-centered things.
- If you are working on a Mac, you already have a screen reader built in! Press command +F5 to try it out yourself. (If not, I know there are other options, just do a quick search).
- Where can you get in napari without using your mouse? Hint: use the tab key to move around major areas and the space bar to interact with elements.

**Comment by goanpeca** [comment]
Thanks for the write up. I will start looking into this. Will need to check the possibilities Qt gives us

Some reference https://doc.qt.io/qt-5/focus.html

**Comment by isabela-pf** [comment]
I got to loop back to this issue for a moment this week and thought it might be helpful to show how the above options look in the light theme (since I mention we are responsible for both light and dark theme). It also may make the issue of ensuring we have contrast in both modes more clear.

All of these have the same specifications as above, repeated so you don't have to scroll between comments. The only difference is the theme they are shown on. 

### Option 1 (light theme)

Color: napari’s main UI blue
Width: 2px
Distance from UI element: 2px

<img width="808" alt="napari preferences with blue focus on interactive elements" src="https://user-images.githubusercontent.com/50221806/140580502-b434f0ba-558b-4bc1-932d-fe1f596f967a.png">


And with an added blue shadow:

<img width="808" alt="napari preferences with blue glowing focus on interactive elements" src="https://user-images.githubusercontent.com/50221806/140580554-7ce4f887-517b-4e05-ba08-0d923b7bc2ee.png">

### Option 2 (light theme)

Color: napari’s main UI blue and white
Width: 2px blue, 1px white
Distance from UI element: 2px

<img width="808" alt="napari preferences with blue and white focus on interactive elements" src="https://user-images.githubusercontent.com/50221806/140580607-e701eef4-cf07-4cc4-bc80-8893a0268b82.png">

### Option 3 (light theme)

Color: napari’s main UI white and dark grey
Width: 2px white, 1px grey
Distance from UI element: 2px

<img width="808" alt="napari preferences with grey and white focus on interactive elements" src="https://user-images.githubusercontent.com/50221806/140580642-4309ef8d-37d8-4bdf-a9ce-adba66b88e3a.png">

---

Based on the [QT Keyboard Focus in Widgets docs](https://doc.qt.io/qt-5/focus.html) (linked above) I'm not more clear on what is or isn't technically feasible, so if anyone has an update on that I'd love to hear it an modify designs to follow suit.

**Comment by dalthviz** [comment]
Hi, I stumbled across this issue while checking #3273 (which seems focus related https://github.com/napari/napari/issues/3273#issuecomment-1618849649) and I think a basic border change could kind of achieve giving a decoration to focused widgets. As an example, if you add a style like the following [in the base style file](https://github.com/napari/napari/blob/main/napari/_qt/qt_resources/styles/00_base.qss):

```
QWidget:focus {
    border: 2px solid {{ current }};
    padding: 2px;
    outline: none;
}
```

You get for the dark theme something like:

![border_decoration](https://github.com/napari/napari/assets/16781833/8e27f866-777b-4333-89bb-034e46039c3e)

Maybe a look to the widgets customization with style sheets and style sheets reference docs could help here:

* https://doc.qt.io/qt-6/stylesheet-customizing.html
* https://doc.qt.io/qt-6/stylesheet-reference.html
